peps/pep-0733.rst

```diff
@@ -0,0 +1,655 @@
+and the particular requirements that each of them has.
+
+
+C API Stakeholders
+==================
+
+As mentioned in the introduction, the C API was originally
+created as the internal interface between CPython's
+interpreter and the Python layer. It was later exposed as
+a way for third-party developers to extend and embed Python
+programs. Over the years, new types of stakeholders emerged,
+with different requirements and areas of focus. This section
+describes this complex state of affairs in terms of the
+actions that different stakeholders need to perform through
+the C API.
+
+Universal Actions for All Stakeholders
+--------------------------------------
+
+There are actions which are generic, and required by
+all types of API users:
+
+* Define functions and call them
+* Define new types
+* Create instances of builtin and user-defined types
+  and perform operations on them
+* Introspect objects, including types, instances, and functions
+* Raise and handle exceptions
+* Import modules
+* Manage threads
+* Manage sub-interpreters
+* Handle and send signals
+
+The following sections look at the unique requirements of various stakeholders.
+
+Extension Writers
+-----------------
+
+Extension writers are the traditional users of the C API. Their requirements
+are the universal actions listed above.
+
+Authors of Embedded Python Applications
+---------------------------------------
+
+Applications with an embedded Python interpreter. Examples are
+`Blender <https://docs.blender.org/api/current/info_overview.html>`__ and
+`OBS <https://obsproject.com/wiki/Getting-Started-With-OBS-Scripting>`__.
+
+They need to be able to:
+
+* Configure the interpreter (import paths, inittab, ``sys.argv``, etc.).
+* Interact with the execution model and program lifetime, including
+  clean interpreter shutdown and restart.
+* Represent complex data models in a way Python can use without
+  having to create deep copies.
+* Provide and import frozen modules.
+* Run multiple independent interpreters (in particular, when embedded
+  in a library that wants to avoid global effects).
+
+Alternative Python Implementations
+----------------------------------
+
+Alternative implementations of Python (such as
+`PyPy <https://www.pypy.org>`__,
+`GraalPy <https://www.graalvm.org/python/>`__,
+`IronPython <https://ironpython.net>`__,
+`RustPython <https://github.com/RustPython/RustPython>`__,
+`MicroPython <https://micropython.org>`__,
+and `Jython <https://www.jython.org>`__), may take
+very different approaches for the implementation of
+different subsystems. They need:
+
+* The API to be abstract and hide implementation details.
+* A specification of the API, ideally with a test suite
+  that ensures compatibility.
+* It would be nice to have an ABI that can be shared
+  across Python implementations.
+
+Alternative APIs
+----------------
+
+There are several projects that implement alternatives to the
+C API, which offer extension users advantanges over programming
```

Comment:
Lumping HPy with `pythoncapi-compat ` and `Argument Clinic` is odd, because they are completely different kinds of things. I doubt the authors of `pythoncapi-compat ` and `Argument Clinic` would call them alternatives to the C API -- they're helpers to be used with the existing C API.

It's likewise awkward to call `Cython` a binding generator. It's more like a whole new language.

Could we perhaps create two sections, one with "API Helpers" (with pythoncapi-compat and Argument Clinic) and another for "Alternative APIs" (with HPy and Cython)? I also wouldn't object to each of HPy and Cython being given their own section. They're both significant efforts to reimagine how extension modules are written.

In the HPy section (or somewhere else) could we also link to the HPy [Manifesto](https://github.com/hpyproject/hpy/wiki/c-api-next-level-manifesto) and its [Why](https://github.com/hpyproject/hpy/wiki/c-api-next-level-changes) document as previous works?

Reply:
See https://github.com/python/peps/pull/3491/files#discussion_r1362383454 re how Cython ended up here (it was previously in the Alternative APIs section).

Reply:
Opening the link doesn't open a specific comment for me, but I did read all of the previous comments and my ones above still stand, I think.

Reply:
Maybe we should just merge "Alternative APIs" and "binding generators"? The needs and concerns are similar, and it's mainly the scope that varies (e.g. Argument Clinic is scoped to parameter conversion, but otherwise it has all the same needs as anything else in this group).

A more interesting distinction could be runtime bindings (e.g. Python.NET) vs. compile-time bindings, but probably that's going to be sufficiently covered by "can't always rely on C macros".